File: array/BinarySearch/boundary/python/BoundarySearch.py
def lower_bound_range_halving(values: list[int], target: int) -> int:
    left = 0
    right = len(values)

    while left < right:
        mid = left + (right - left) // 2
        if values[mid] < target:
            left = mid + 1
        else:
            right = mid

    return left


def lower_bound_powers_of_two(values: list[int], target: int) -> int:
    position = -1
    step = 1
    while step < len(values):
        step <<= 1

    while step > 0:
        next_index = position + step
        if next_index < len(values) and values[next_index] < target:
            position = next_index
        step >>= 1

    return position + 1

File: array/BinarySearch/boundary/cpp/BoundarySearch.cpp
#include "BoundarySearch.hpp"

#include <cstddef>

int LowerBoundRangeHalving(const std::vector<long long>& values, long long target) {
    int left = 0;
    int right = static_cast<int>(values.size());

    while (left < right) {
        const int mid = left + (right - left) / 2;
        if (values[mid] < target) {
            left = mid + 1;
        } else {
            right = mid;
        }
    }

    return left;
}

int LowerBoundPowersOfTwo(const std::vector<long long>& values, long long target) {
    int position = -1;
    int step = 1;
    while (step < static_cast<int>(values.size())) {
        step <<= 1;
    }

    for (; step > 0; step >>= 1) {
        const int next = position + step;
        if (next < static_cast<int>(values.size()) && values[next] < target) {
            position = next;
        }
    }

    return position + 1;
}
